### ISSUE TITLE
Assets slowness in displaying on screen

### ISSUE THREAD

**Issue by cepacs** [snipe/snipe-it#5511]
#### Expected Behavior (or desired behavior if a feature request)

Click on "All Ready to Deploy" and have the assets display in a reasonable time..

-----

#### Actual Behavior

When we click on "All Ready to Deploy" it takes 3min 30sec to populate the screen, which is displaying 1-100 of 9269 assets.  If I click on page two it takes another 3min to display 101-200.

If I connect with SQL or Access and open the tables they do respond very quickly.
We took v4.3.0 - build 3559 (master) last week and I am wondering if that had anything to do with it or after an update is there something we need to run to optimize the DB\Web interface at all?  

-----

#### Please confirm you have done the following before posting your bug report:

- [X ] I have enabled debug mode 
- [X ] I have read [checked the Common Issues page](https://snipe-it.readme.io/docs/common-issues)

-----
#### Provide answers to these questions:

- Is this a fresh install or an upgrade? **upgrade**
- Version of Snipe-IT you're running **Version v4.3.0 build 3559 (master)**
- Version of PHP you're running **PHP version: 7.1.16**
- Version of MySQL/MariaDB you're running **MySQL Server version: 5.7.21**
- What OS and web server you're running Snipe-IT on **Windows Server 2016 / IIS**
- What method you used to install Snipe-IT (install.sh, manual installation, docker, etc) **GIT**
- WITH DEBUG TURNED ON, if you're getting an error in your browser, include that error
- What specific Snipe-IT page you're on, and what specific element you're interacting with to trigger the error
- If a stacktrace is provided in the error, include that too.
- Any errors that appear in your browser's error console.
- Confirm whether the error is reproducible on the demo: https://snipeitapp.com/demo.
- Include any additional information you can find in `storage/logs` and your webserver's logs.
- Include what you've done so far in the installation, and if you got any error messages along the way.
- Indicate whether or not you've manually edited any data directly in the database

Please do not post an issue without answering the related questions above. If you have opened a different issue and already answered these questions, answer them again, once for every ticket. It will be next to impossible for us to help you.

https://snipe-it.readme.io/docs/getting-help

**Comment by EarlRamirez** [comment]
@cepacs I have approximately the same number of assets and I do not see this behaviour, what are the resources your server is running. Also, you mentioned if you access the tables it respond quickly, are you running the same query that the UI is running, with join etc.?

**Comment by cepacs** [comment]
I hadn't tried replicating the query it is passing within SQL.
When it is loading the screen it is only using less than 2GB of the 16GB assigned.  The processor (2 @ 2.4GHz) spikes to about 50% for the duration until results are shown. Ethernet shows almost no activity 8Kbps Spikes
We are running IIS on Server 2016, MySQL 5.7.21 ,PHP 7.1.17

I did notice that if I click on All Archived I do get the results quick, probably because I only have 122 assets.  So it seems like it is able to process the whole query quickly but displaying takes longer as you scale up in results..

**Comment by pcsmarts502** [comment]
I am getting the same results as you.  If I click on pending which has about 20 records it pulls up fast.  MySQL spikes similar to what you have described.  There are no errors in the logs.  I am not sure where to search next.

**Comment by pcsmarts502** [comment]
Previous versions of snipe-it uses this query to list all.
SELECT 
    `assets`.*
FROM
    `assets`
        INNER JOIN
    `status_labels` AS `status_alias` ON `status_alias`.`id` = `assets`.`status_id`
        AND `status_alias`.`archived` = 0
WHERE
    `assets`.`deleted_at` IS NULL
ORDER BY `name` ASC
LIMIT 500 OFFSET 0

Any Idea what happened here?

**Comment by cepacs** [comment]
That is strange.. the previous one does look like it would return much faster than the newer version.  Maybe it can be optimized?  I wish I had more time to look at this.. :/

**Comment by snipe** [comment]
> I have the same problem with 4.3. I have been trying to figure it out for weeks now with no luck.

That seems unlikely, as 4.3.0 was just released 8 days ago.

> We need to keep this thread alive or we will be stuck on 4.2.0

That's a bit much, really, but okay. Threads auto-close after months of inactivity. Calm down. We're looking at this issue, trying to replicate it.

I'm testing now by rolling back the counts that we added. It may be a thing we need to de-normalize (which sucks, but..)

I assume you folks have tried upping your MySQL and PHP memory allocations?

**Comment by sid8000** [comment]
> I assume you folks have tried upping your MySQL and PHP memory allocations?

I have tried allocating more memory but still the same. 

In our inventory there are only 1000 assets still it takes 25-30 secs to load a page irrespective of number of rows per page.

**Comment by cepacs** [comment]
Yes.. thanks for looking into this.  We had adjusted our memory allocations to 1G in PHP and MySQL just to see if anything would change, but load times did stay consistent at about 3 and half minutes for about 9000 records.

**Comment by snipe** [comment]
> In our inventory there are only 1000 assets still it takes 25-30 secs to load a page irrespective of number of rows per page

That's bonkers. 

>  load times did stay consistent at about 3 and half minutes for about 9000 records.

Also bonkers.

Is there anything in common for all of you? Tons of locations, tons of users, anything? I'm testing locally with 15k assets and it's a *little* slower than it should be, but not minutes-worth. That query is doing a LOT of work, but with only 9k assets, MySQL shouldn't really feel that.

Have you tried turning query caching on? Can you give me info on your memory configs? Do you have any caching modules like APC or OpCache enabled?

**Comment by sid8000** [comment]
In my case, 5 locations, around 500 users and 1K assets. Strange thing is this slowness occured when upgraded to 4.3.0

**Comment by snipe** [comment]
@sid8000 we did add counters, which taps into the action log. If your action_logs table has a lot of entries, that might be part of the issue. I'm exploring possible solutions (adding an additional index on the action_logs table for action_type) or de-norming the counters. Hard for me to troubleshoot since I can't reproduce tho :(

**Comment by sid8000** [comment]
Yes. Action logs has too many entries. Is there any way to limit that? Since all the action log is reflecting in history.

**Comment by sid8000** [comment]
At least 15-17K.......

**Comment by cepacs** [comment]
I have 9400 assets and 19250 in action_logs.  The numbers don't seem very large to me honestly but maybe it depends on how it is executing the query.  If it has to query all 19K records for every one asset each time then I could see that adding up.

**Comment by snipe** [comment]
Yeah, none of those numbers should make MySQL even bat an eyelash - but I think adding an index could help here, preventing a full table scan each time. 

Either that, or we de-norm those values. De-norming is less desirable, just because it means we need to change a bunch of code for all of the checkins/checkouts/requests/canceled requests, but I'm still digging around to see where our best performance gains can be had.

**Comment by cepacs** [comment]
De-normalizing shouldn't be needed I wouldn't think.  I think indexing would make a huge difference.. :+1:

**Comment by pcsmarts502** [comment]
Still loving this app.  Is it possible to take this part out of the query thats being generate?

        INNER JOIN
    `status_labels` AS `status_alias` ON `status_alias`.`id` = `assets`.`status_id`
        AND `status_alias`.`deployable` = 1
        AND `status_alias`.`pending` = 0
        AND `status_alias`.`archived` = 0

When  I test it without the INNER JOIN it works perfectly.  This is for the list all and ready to deploy APIs

**Comment by adambowen** [comment]
We're experiencing the same issue, I've simply disabled the counts in the fork we use. Previously it was taking over 3 minutes to load the asset list, without them it takes around 600ms.

In case it helps, we have ~6500 assets, and ~21000 action logs.

We use Postgres for the products we develop, so I'm not terribly familiar with MySQL performance - however, looking at the query posted earlier I'd suggest an index on `action_logs.item_id` or `(action_logs.item_id, action_logs.item_type)` would help a lot - it looks like it might do a full table scan of action logs for every asset currently.

```
EXPLAIN SELECT assets.*, (SELECT COUNT(*) FROM action_logs WHERE assets.id = action_logs.item_id AND item_type = 'App\Models\Asset' AND action_type = 'checkin from') AS checkins_count, (SELECT COUNT(*) FROM action_logs WHERE assets.id = action_logs.item_id AND item_type = 'App\Models\Asset' AND action_type = 'checkout') AS checkouts_count, (SELECT COUNT(*) FROM action_logs WHERE assets.id = action_logs.item_id AND item_type = 'App\Models\Asset' AND action_type = 'requested') AS user_requests_count FROM assets INNER JOIN status_labels AS status_alias ON status_alias.id = assets.status_id AND status_alias.archived = 0 WHERE asse
+----+--------------------+--------------+------------+------+---------------+------+---------+------+-------+----------+----------------------------------------------------+
| id | select_type        | table        | partitions | type | possible_keys | key  | key_len | ref  | rows  | filtered | Extra                                              |
+----+--------------------+--------------+------------+------+---------------+------+---------+------+-------+----------+----------------------------------------------------+
|  1 | PRIMARY            | status_alias | NULL       | ALL  | PRIMARY       | NULL | NULL    | NULL |    13 |    10.00 | Using where; Using temporary; Using filesort       |
|  1 | PRIMARY            | assets       | NULL       | ALL  | NULL          | NULL | NULL    | NULL |  6504 |     1.00 | Using where; Using join buffer (Block Nested Loop) |
|  4 | DEPENDENT SUBQUERY | action_logs  | NULL       | ALL  | NULL          | NULL | NULL    | NULL | 20792 |     0.10 | Using where                                        |
|  3 | DEPENDENT SUBQUERY | action_logs  | NULL       | ALL  | NULL          | NULL | NULL    | NULL | 20792 |     0.10 | Using where                                        |
|  2 | DEPENDENT SUBQUERY | action_logs  | NULL       | ALL  | NULL          | NULL | NULL    | NULL | 20792 |     0.10 | Using where                                        |
+----+--------------------+--------------+------------+------+---------------+------+---------+------+-------+----------+----------------------------------------------------+
5 rows in set, 4 warnings (0.00 sec)
```

**Comment by adambowen** [comment]
I figured I'd try adding the index...

```
mysql> create index action_logs_item_id_item_type ON action_logs(item_id, item_type);                                                                                                                              Query OK, 0 rows affected (0.08 sec)
Records: 0  Duplicates: 0  Warnings: 0
```
```
mysql> EXPLAIN SELECT assets.*, 
(SELECT COUNT(*) FROM action_logs WHERE assets.id = action_logs.item_id AND item_type = 'App\Models\Asset' AND action_type = 'checkin from') AS checkins_count, 
(SELECT COUNT(*) FROM action_logs WHERE assets.id = action_logs.item_id AND item_type = 'App\Models\Asset' AND action_type = 'checkout') AS checkouts_count, 
(SELECT COUNT(*) FROM action_logs WHERE assets.id = action_logs.item_id AND item_type = 'App\Models\Asset' AND action_type = 'requested') AS user_requests_count 
FROM assets 
INNER JOIN status_labels AS status_alias ON status_alias.id = assets.status_id AND status_alias.archived = 0 
WHERE assets.deleted_at IS NULL ORDER BY name ASC LIMIT 20 OFFSET 0;
+----+--------------------+--------------+------------+------+-------------------------------+-------------------------------+---------+-------------------------+------+----------+----------------------------------------------------+
| id | select_type        | table        | partitions | type | possible_keys                 | key                           | key_len | ref                     | rows | filtered | Extra                                              |
+----+--------------------+--------------+------------+------+-------------------------------+-------------------------------+---------+-------------------------+------+----------+----------------------------------------------------+
|  1 | PRIMARY            | status_alias | NULL       | ALL  | PRIMARY                       | NULL                          | NULL    | NULL                    |   13 |    10.00 | Using where; Using temporary; Using filesort       |
|  1 | PRIMARY            | assets       | NULL       | ALL  | NULL                          | NULL                          | NULL    | NULL                    | 6504 |     1.00 | Using where; Using join buffer (Block Nested Loop) |
|  4 | DEPENDENT SUBQUERY | action_logs  | NULL       | ref  | action_logs_item_id_item_type | action_logs_item_id_item_type | 770     | snipeit.assets.id,const |    3 |    10.00 | Using index condition; Using where                 |
|  3 | DEPENDENT SUBQUERY | action_logs  | NULL       | ref  | action_logs_item_id_item_type | action_logs_item_id_item_type | 770     | snipeit.assets.id,const |    3 |    10.00 | Using index condition; Using where                 |
|  2 | DEPENDENT SUBQUERY | action_logs  | NULL       | ref  | action_logs_item_id_item_type | action_logs_item_id_item_type | 770     | snipeit.assets.id,const |    3 |    10.00 | Using index condition; Using where                 |
+----+--------------------+--------------+------------+------+-------------------------------+-------------------------------+---------+-------------------------+------+----------+----------------------------------------------------+
5 rows in set, 4 warnings (0.00 sec)
```

The query is now nearly instant through a mysql prompt (previously it took longer than I was willing to wait).

**Comment by sid8000** [comment]
> create index action_logs_item_id_item_type ON action_logs(item_id, item_type);  
Yup. That's working. Instant results even on default MySQL configuration.
Thanks Adam.

**Comment by snipe** [comment]
Hey, so, you should be aware that when you manually add indexes, you're likely going to run into errors in the migrations later down the line, as we add the same indexes through the normal upgrade process.

We were testing with the following indexes:

```
Schema::table('action_logs', function (Blueprint $table) {
            $table->index(['target_id', 'target_type']);
            $table->index('created_at');
            $table->index(['item_type', 'item_id', 'action_type']);
            $table->index(['target_type', 'target_id', 'action_type']);
        });
```

And although it improves performance, it's not as dramatic as what you're seeing.

**Comment by snipe** [comment]
> explain select `assets`.*, (select count(*) from `action_logs` where `assets`.`id` = `action_logs`.`item_id` and `item_type` = 'App\Models\Asset' and `action_type` = 'checkin from') as `checkins_count`, (select count(*) from `action_logs` where `assets`.`id` = `action_logs`.`item_id` and `item_type` = 'App\Models\Asset' and `action_type` = 'checkout') as `checkouts_count`, (select count(*) from `action_logs` where `assets`.`id` = `action_logs`.`item_id` and `item_type` = 'App\Models\Asset' and `action_type` = 'requested') as `user_requests_count` from `assets` inner join `status_labels` as `status_alias` on `status_alias`.`id` = `assets`.`status_id` and `status_alias`.`archived` = '0' where `assets`.`deleted_at` is null order by `id` desc limit 200 offset 400

<img width="1316" alt="screen shot 2018-05-16 at 12 57 12 pm" src="https://user-images.githubusercontent.com/197404/40140852-c8cc207c-5908-11e8-901d-a38276592f1d.png">

The reason why this is so complicated is because there are a ton of different sorting options, which makes those indexes and queries pretty complex.

**Comment by snipe** [comment]
I have a PR on develop right now if anyone wants to take a look. (Backup your database before trying this, of course). Could use some help kicking the tires on some higher volume users (lots of assets, lots of action_logs, or both.)

**Comment by snipe** [comment]
Can anyone confirm that this has resolved their issue?

**Comment by sid8000** [comment]
> create index action_logs_item_id_item_type ON action_logs(item_id, item_type);

This works for me.
I have tried adding action_type in above query but it results in error 1071 : Specified key was too long: max key length is 1000 bytes.
I don't know if it is the right way to do it or not.

**Comment by cepacs** [comment]
I took the the develop update and I am still seeing dramatic results when clicking on all assets.. less than 2 seconds to populate for me.. I have assets (9400 records) and action_logs (19400 records).  

Feels like a win to me.. :)

Question.. Can I stay on this develop branch and then just upgrade back to a master branch or is that not a good idea? 

Thanks for staying on this issue.. Long live #5511!!  lol ;P

**Comment by snipe** [comment]
@sid8000 Once again, you should not be creating keys manually. They ca (and often do) conflict with keys we add, and then you have to delete them before you can complete migrations.

@cepacs great news - it's actually already on master now (we're prepping for an official release), so you should be able to switch back to master and have no issues.

**Comment by sid8000** [comment]
@snipe Yes. All which I have tried is on my testing machine. No experiments on live data.
But now I have tried to update to latest version on testing machine, previous database deleted, created new db restored the from live machine's db. It gives same error. Created new issue.
#5575

**Comment by sid8000** [comment]
Found out the solution. Tables are created using MYISAM engine. Converted action_logs table to InnoDB. Migration successful. All these experiments are on testing machine.
(Is there any side effects of changing storage engine that I should worry about?)
And also I do observed one thing @adambowen 's query resulted in way too fast while displaying all assets compared to what results are getting in 4.4.1

**Comment by snipe** [comment]
There isn’t, because we create the tables as InnoDB. I can’t imagine why yours were created as MYISAM. That hasn’t been the default for MySQL in years.

**Comment by snipe** [comment]
This seems to be an issue for some other users (myisam vs innodb) and i’m genuinely baffled as to how it’s an issue. InnoDB has been the default in MySQL for years and years, plus we specify that in the DB config. Perplexing, but I appreciate the extra insight here. I’ll figure out a fix.

**Comment by snipe** [comment]
Turns out, we don't specify the engine anymore. I might have just overlooked it during a framework upgrade, or I may have taken it out since InnoDB has been the default since 5.6, I'm honestly not sure. 

I've added the default engine to the DB configs, and I guess I'll just have to add some docs about how to manually switch engines for the folks who are for some reason still using MyISAM as a default.

**Comment by sid8000** [comment]
I don't why WAMPSERVER using MYISAM as default engine then.
![image](https://user-images.githubusercontent.com/30995759/40403827-00085e52-5e72-11e8-9dc8-7ec245b84168.png)
